coalib/core/PersistentHash.py

```diff
@@ -1,19 +1,42 @@
+from collections import Iterable
 from hashlib import sha1
 import pickle
 
 
-def persistent_hash(obj):
+def order(obj):
+    """
+    Recursively preserves the order of any sets or dicts
+    found in obj.
     """
-    Calculates a persistent hash of the given object.
+    if isinstance(obj, set):
+        return sorted(obj)
+    if isinstance(obj, dict):
+        temp_obj = obj
+        for k, v in temp_obj.items():
+            if isinstance(v, set) or isinstance(v, dict):
+                temp_obj[k] = order(v)
+        return sorted(temp_obj.items())
+    return obj
 
-    This hash method uses pickle (protocol 4) to serialize the given object and
-    hash the byte-stream. The hashing algorithm used is SHA-1.
 
-    :param obj:
-        The object to calculate the persistent hash for.
-    :return:
-        The persistent hash.
-    """
+def preserve_order(a, k):
+    arg_list = []
+    kwarg_dict = {}
+
+    for ele in a:
+        arg_list.append(order(ele))
```

Comment:
```python
args = tuple(order(ele))
kwargs = sorted({key: order(value) for key, value in k.items()})

return args, kwargs
```

Reply:
I tried it but it gives me a wrong hash.

Reply:
I have made the changes to the kwargs code but kept the args one as it is.